### PR TITLE
feat(GroupChats): Implemented group admin permissions

### LINKED ISSUE #5941: Group Chat - Add / remove members - New spec
### Description
New spec:
1) When a user starts creating a new 1 on 1 or ad-hoc group chat, they should be shown at the beginning (left) of the add user pill input field, using the 'king pill'.  In this state there is no "X" on the 'king pill' because it doesn't make any sense for a user to remove themself from a group chat before they have even started it.
2) When a user who is not an admin enters the 'edit mode', all users who are already in the chat are represented by grey pillis without an "X".
3) When the admin enters the edit mode, there are "x"'s on all pills excluding their own 'king pill'.
4) Everybody in the ad-hoc group chat who is not the admin will see a "Add to Group" menu option instead of an "Add / remove from group" menu option (only the ad-hoc group chat admin will see this in the menu).  And when someone who is not the admin selects the "Add to Group" menu option, they shouldn't see "X" on the right of the pills belonging to users who are already in the ad-hoc group chat.

image

Related to https://github.com/status-im/status-desktop/issues/5522
Needs implementation of https://github.com/status-im/StatusQ/issues/694

### PR DESCRIPTION
**Pull request by alexandraB99**
Closes #5941
Depends on: https://github.com/status-im/StatusQ/pull/860

### What does the PR do
Implements group admin permissions

### Affected areas
Group chats

### Screenshot of functionality (including design for compari
https://user-images.githubusercontent.com/31625338/185178867-98f09194-f930-4b48-9905-ae24d59cea61.mov

### CODE REVIEW

**Review by noeliaSD — COMMENTED** [review]
Pending review until we agree on this PR https://github.com/status-im/StatusQ/pull/860

**Comment by PascalPrecht** [comment]
This looks really cool!
Only thing that jumps straight into my face is that the scrollable area inside of the tag list seems to be too small. There's a lot of empty space left in the input that could be used to render the tags.

File: ui/app/AppLayouts/Chat/panels/GroupChatPanel.qml [hunk 1]
```diff
@@ -76,24 +82,30 @@ RowLayout {
...
+                 d.contactsModel.insert(d.contactsModel.count,
+                                        {"pubKey": entry2.pubKey,
+                                         "displayName": entry2.name,
+                                         "localNickname": !!entry2.localNickname ? entry2.localNickname : "",
+                                         "icon": entry2.icon,
+                                         "isAdmin": entry.isAdmin,
+                                         "isReadonly": !root.rootStore.isCurrentUser(entry.pubKey),
+                                         "isVerified": !!entry2.isVerified ? entry2.isVerified : false,
+                                         "isUntrustworthy": !!entry2.isUntrustworthy ? entry2.isUntrustworthy : false,
+                                         "isContact": !!entry2.isContact ? entry2.isContact : false,
```
Comment (endulab) [inline ui/app/AppLayouts/Chat/panels/GroupChatPanel.qml:106]: Why these !! checks are needed. May model not contain this data ?
If no maybe it's better to do checks at the beginning when filling item (line 67 ....) ?

**Comment by glitchminer** [comment]
## Issues Found

---
### 1. OPEN When starting chat pill is missing remove button if selected from dropdown
Actual


---
### 2. OPEN On To field, pressing enter with the dropdown open creates a pill with no contact
Actual


---
### 3. OPEN When multiple contacts are displayed in drop-down it is not possible to use arrow keys to select one
Actual
image

---
### 4. OPEN After clicking on "Add to group" 1 of 2 non contacts are shown using username instead of display name
Actual
image

---
### 5. OPEN After clicking on "Add to group" 2 contacts are shown with the X to remove icon in pill
Actual
image

---
### 6. OPEN After clicking on "Add to group" the pills are blue despite them not being editable
Actual
image

---
### 7. OPEN At minimum width it is not possible to view/edit all users in the field with either keyboard or mouse
Actual
image

---
### 8. OPEN There is a large margin on the right hand side of the field (even when max user label is not shown)
Actual
image

---
### 9. OPEN Dropdown does not always align under current cursor position
Actual
image

---
### 10. OPEN 5 user limit is not enforced
Expected
image

Actual
image


---
### Other observations:
- It is not possible to right click to paste into the To field
- It is not possible to highlight text in the To field with a mouse
- After adding or removing users the default group name may no longer be accurate

**Review by glitchminer — CHANGES_REQUESTED** [review]
Issues found and noted in comment https://github.com/status-im/status-desktop/pull/7057#issuecomment-1229236321

**Comment by alexandraB99** [comment]
> This looks really cool! Only thing that jumps straight into my face is that the scrollable area inside of the tag list seems to be too small. There's a lot of empty space left in the input that could be used to render the tags.

@PascalPrecht could you please open a new issue for that? I think @glitchminer is mentioning it as well.

**Comment by alexandraB99** [comment]
@glitchminer issues 5,6 should be fixed. Issues 2,3, 7, 8, 9, 10 are not relevant to https://github.com/status-im/status-desktop/issues/5941, so could you please open new issues for these? Issue 1: there is no requirement that admin is on the names list when opening to create a new group chat see https://www.figma.com/file/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?node-id=391%3A99912. Issue 4, please check again it might be that not all users have set a display name, in this case the username appears.

File: ui/app/AppLayouts/Chat/panels/GroupChatPanel.qml (hunk 1 above)
Comment (alexandraB99) [reply ui/app/AppLayouts/Chat/panels/GroupChatPanel.qml]: removed

**Comment by glitchminer** [comment]
> 

Hi @alexandraB99, retested and issue 5 and 6 appear unresolved. A user who is not admin is still seeing other users in blue and is still given the cross to remove. If it's not possible to resolve in this one, the missing display names can be moved with issue 4. 
image

File: ui/app/AppLayouts/Chat/panels/GroupChatPanel.qml [hunk 2]
```diff
@@ -76,24 +82,29 @@ RowLayout {
...
          // Build contactsModel type from model type (to fit with expected StatusTagSelector format
          for (var j = 0; j < contactsModelListView.count; j ++) {
              var entry2 = contactsModelListView.itemAtIndex(j)
              d.contactsModel.insert(d.contactsModel.count,
-                                {"pubKey": entry2.pubKey,
-                                 "displayName": entry2.name,
-                                 "icon": entry2.icon,
-                                 "isIdenticon": false,
-                                 "onlineStatus": false})
+                                    {"pubKey": entry2.pubKey,
```
Comment (micieslak) [inline ui/app/AppLayouts/Chat/panels/GroupChatPanel.qml:97]: I know it's only matter of style, but there is no need to wrap js dict keys in "", imo it looks cleaner without quotation marks.
Other minor comment: insert could be replaced by append.

**Comment by glitchminer** [comment]
@alexandraB99, per comment, I've created and moved the following issues: 
  2. https://github.com/status-im/status-desktop/issues/7234
  3. https://github.com/status-im/status-desktop/issues/7235
  4. https://github.com/status-im/status-desktop/issues/7236
  7. https://github.com/status-im/status-desktop/issues/7237
  8. https://github.com/status-im/status-desktop/issues/7238
  9. https://github.com/status-im/status-desktop/issues/7239
  10. https://github.com/status-im/status-desktop/issues/7240

1, 5 and 6 remain open. I know you mentioned issues 1 above but it's actually regarding the close button not appearing when it should, not about the admin user.

**Comment by PascalPrecht** [comment]
@alexandraB99 Looks like @glitchminer beat me to it: https://github.com/status-im/status-desktop/issues/7238

File: ui/app/AppLayouts/Chat/panels/GroupChatPanel.qml (hunk 2 above)
Comment (alexandraB99) [reply ui/app/AppLayouts/Chat/panels/GroupChatPanel.qml]: done!

**Comment by alexandraB99** [comment]
> > 
> 
> Hi @alexandraB99, retested and issue 5 and 6 appear unresolved. A user who is not admin is still seeing other users in blue and is still given the cross to remove. If it's not possible to resolve in this one, the missing display names can be moved with issue 4. image

@glitchminer thanks for re-testing! issues 1, 5 and 6 should be fixed now. As about the empty display names is it reported in a separate issue?

**Comment by glitchminer** [comment]
Hi @alexandraB99, regarding 5 & 6 is it expected that a user should be able to remove themselves from the group via this mechanism?
image

Issue 1 is now closed. 

Empty display names are covered by https://github.com/status-im/status-desktop/issues/7236

**Comment by alexandraB99** [comment]
> Hi @alexandraB99, regarding 5 & 6 is it expected that a user should be able to remove themselves from the group via this mechanism? image
> 
> Issue 1 is now closed.
> 
> Empty display names are covered by #7236

@glitchminer I think yes, from the requirements in the [issue](https://github.com/status-im/status-desktop/issues/5941):

> When a user who is not an admin enters the 'edit mode', all users who are already in the chat are represented by grey pillis without an "X".
only the other users shouldn't have a "x".

**Comment by glitchminer** [comment]
> @glitchminer I think yes, from the requirements in the [issue](https://github.com/status-im/status-desktop/issues/5941):
> 
> > When a user who is not an admin enters the 'edit mode', all users who are already in the chat are represented by grey pillis without an "X".
> 
> only the other users shouldn't have a "x".

Thanks @alexandraB99, looking at that now. To me, it doesn't specifically say that the non-admin user has a separate behavior from "all users..". I'd expect them all to be grey. Also, the current behaviour when a non-admin user removes themselves like this is different from "Leave group" function.
- Leave group: Group chat immediately closes and is removed from sidebar
- Remove button on pill: Group chat remains open with previous messages visible

**Comment by John-44** [comment]
> 

In an ad-hoc group chat, there is no such thing as 'all users', a user is either an 'admin of the ad-hoc group chat' or 'not an admin of the ad-hoc group chat'.

When a user who is 'not an admin of the ad-hoc group chat' opens this edit mode, they can add new users to the ad-hoc group chat (as long as those users are mutual contacts, and as long as the total number of people in the ad-hoc group chat does not exceed the max number of participants).  

This means that when a user who is 'not an admin of the ad-hoc group chat' opens this edit mode, all existing members of the ad-hoc group chat should have grey pills without an 'X'.  When the user who is 'not an admin of the ad-hoc group chat' adds more users, they should be displayed as blue pills without an 'X' *until* the user hits 'Confirm'.  If after adding users and then exiting the edit mode, the same user who is 'not an admin of the ad-hoc group chat' opens edit mode again, they should see *all* current members of the ad-hoc group chat (including those that user added previously) displayed as grey pills without an 'X'.

This is because users who are not 'not an admin of the ad-hoc group chat' can add mutual contacts to the group chat, but only the admin of that ad-hoc group chat can remove users after they are added.

@alexandraB99 does this answer your question?  cc'ing @osmaczko

**Comment by alexandraB99** [comment]
> @alexandraB99 there is a [PR](https://github.com/status-im/status-desktop/pull/7188) that addresses [all the issues](https://github.com/status-im/status-desktop/pull/7057#issuecomment-1237412114) related to members selector. I have removed `GroupChatPanel` because it is unmaintainable with its imperative filtering and sorting. Also if you take a look at the console there are a lot of warnings from `StatusTagSelector` because of incompatible models - the names model is not compatible with the delegate that expects a complete user model with properties like displayName, localNickname, isContact, etc.
> 
> I am not sure if you should focus on finalizing this PR because `GroupChatPanel` will get removed eventually. Also mentioned PR covers task you are working on.

@osmaczko good to know, closing this PR as it's not relevant anymore.

**Comment by alexandraB99** [comment]
> @alexandraB99 there is a [PR](https://github.com/status-im/status-desktop/pull/7188) that addresses [all the issues](https://github.com/status-im/status-desktop/pull/7057#issuecomment-1237412114) related to members selector. I have removed `GroupChatPanel` because it is unmaintainable with its imperative filtering and sorting. Also if you take a look at the console there are a lot of warnings from `StatusTagSelector` because of incompatible models - the names model is not compatible with the delegate that expects a complete user model with properties like displayName, localNickname, isContact, etc.
> 
> I am not sure if you should focus on finalizing this PR because `GroupChatPanel` will get removed eventually. Also mentioned PR covers task you are working on.

@osmaczko good to know, closing this PR as it's not relevant anymore.

**Pull request closed without merging**